Reduce ABM checks.

lhofhansl opened the pull request:
See #7280

Since is simple and low risk change, reducing the number of time blocks needs to be checked for whether ABMs needs to be run or not, by doing some simple caching.

This builds a cache of the node types in a blocks (up to ~~64~~ 32) and stores this with a cached block. Before any ABMs are checked we first check against this set.

The advantage is that cache invalidation is simple and only depends on a single block; the risk of incorrectly handling this is low.
The disadvantage is that we lose out of further optimization - for example we could cache whether the neighbor check failed, but then we'd need a more tricky protocol for cache invalidation.

Have a look. Something to discuss; but it is functioning as is.

lhofhansl commented:
Hmm... Doesn't quite work with active objects...? They seem to pop in and out of existence.

lhofhansl commented:
Some minor improvements... Flew around in a few worlds, did not find a single block with more than 32 different content_types, so just caching 32 now (1/128 of the possible 4096 content types).
Also add some better stats.

Example: Just with the ethereal mod, `active_object_range` set to 7. I.e. 2084 active blocks, just 87 of those need to actually be checked:
![screenshot_20180715_035236](https://user-images.githubusercontent.com/446130/42733260-36172510-87e3-11e8-8fe3-1de3668b9224.png)

That's a pretty significant saving... I did not see the 200ms warning at all.

lhofhansl commented:
More complex world with all kinds of mods (including NSSM) with lots of ABMs.

Still just 528 of the 2103 active blocks even need to be scanned.

![screenshot_20180715_040143](https://user-images.githubusercontent.com/446130/42733292-b9b7dc66-87e3-11e8-9cac-0e70bd26c5c5.png)

nerzhul commented on src/mapblock.h:
```diff
@@ -519,6 +521,9 @@ class MapBlock
 #ifndef SERVER // Only on client
 	MapBlockMesh *mesh = nullptr;
 #endif
+	std::set<content_t> contents;
```
unordered_set ?

lhofhansl replied:
Oops. Yes.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -800,11 +800,29 @@ class ABMHandler
 
+		if (block->contents_cached) {
+			bool run_abms = false;
+			for(content_t c : block->contents) {
```
missing space

ClobberXD replied:
on L805 too

nerzhul commented:
good design, it's the design i wanted to add to mapblocks at a moment but didn't get time for it.

lhofhansl commented:
Addressed comments.

Thanks on the design. :)
I'm still thinking about how to effectively cache the neighbor checks without doing a lot of work at write time and/or risking subtle bugs in the cache invalidation.

lhofhansl commented:
I think if/when this is merged we can increase the default active_block_range to 4... (Although I expect some opposition from @paramat :) )

lhofhansl commented:
With the "plantlife" modpack I find some blocks with more than 32 content types (4% of the blocks). With 64 is back down to 0.

So probably the 64 types limit was better. Or maybe it's not worth it...?
We're talking about 64 shorts, i.e. 128 bytes vs 64 bytes per cached block.

Fixer-007 commented:
Try this world save too https://github.com/minetest/minetest/files/101456/Just.Test.zip

lhofhansl commented:
From the Just.Test world (I hope you didn't sneak any malicious lua scripts into that one).
Looks like 64 is still a good number...

![screenshot_20180715_070309](https://user-images.githubusercontent.com/446130/42734678-0e64cd78-87fd-11e8-8f6b-7e27d59bf378.png)

paramat commented:
> I think if/when this is merged we can increase the default active_block_range to 4

I doubt it, because there will be worlds with complex built-up areas where almost all blocks are over the cache size so all blocks have ABMs running the more intensive way, these are also the areas where ABMs are working the hardest and triggering the most lua actions.
The result is that you will completely lose any performance gained from this optimisation and cause much worse performance in these areas. Better to keep range 3 and have a significant improvement in ABM performance everywhere.

My initial concern is the memory use of storing the caches, how large is a cache compared to a mapblock, i guess it's 64 added to 4096 * 2?
Cache invalidation is lightweight i assume because the cache is already loaded?

paramat commented:
So these caches are not saved to world DB? Or are they?

lhofhansl commented:
Nope. Not saved, cached the first time we try to execute ABMs. Hardly any extra work, nor any extra storage. About 140 bytes extra in RAM per cached block.

paramat commented:
Ok that's fine by me.

The worst case scenario is 1 player walking constantly through a world. With active block range 3, a certain block will be active for (48 * 2 + 16) / 4 = 28s so the block will be cached in the first second and the optimisation effective for the following 27s. Seems good.

sfan5 commented on src/serverenvironment.cpp:
```diff
@@ -800,11 +800,29 @@ class ABMHandler
+			}
+			if (!run_abms) {
+				return;
+			}
```
brace removal

sfan5 commented on src/serverenvironment.cpp:
```diff
@@ -800,11 +800,29 @@ class ABMHandler
 
+		if (block->contents_cached) {
+			bool run_abms = false;
+			for (content_t c : block->contents) {
```
this could use a comment to easily see what it does

lhofhansl replied:
Yep!

sfan5 commented on src/serverenvironment.cpp:
```diff
@@ -818,6 +836,14 @@ class ABMHandler
 			content_t c = n.getContent();
+			if (!block->contents_cached && !block->do_not_cache_contents) {
+				block->contents.insert(c);
+				if (block->contents.size() > 32) {
```
does it make sense for this to be configurable?

lhofhansl replied:
IMHO. We should find a good default. 64 turns out to be almost universally good.

lhofhansl commented:
With default settings and a mostly default world about 6ms are spent in ABMs without this PR, and about 1ms with this PR.
If I crank up active_object_range to 13, it's about 120-130ms without this PR, and about 30ms with this PR.

lhofhansl commented:
I will say that that was with a single player only (me). With default settings and 10 players we're talking 60ms just for the scanning when mostly nothing happens (vs less than 10ms with this PR).
So I think it's definitely worth it.

lhofhansl commented:
Let's get this in...?

rubenwardy commented:
The code looks good to me on inspection, but I'm not particularly comfortable with this area of code

SmallJoker commented on src/mapblock.h:
```diff
@@ -519,6 +521,9 @@ class MapBlock
 #endif
+	std::unordered_set<content_t> contents;
+	bool contents_cached = false;
+	bool do_not_cache_contents = false;
```
The variables should state that they're only for ABMs and not helpful in any other area.

SmallJoker commented on src/serverenvironment.cpp:
```diff
@@ -1311,8 +1345,12 @@ void ServerEnvironment::step(float dtime)
+				abmhandler.apply(block, abm_checked, abm_run, abm_not_cached);
 			}
+			g_profiler->avg("SEnv: active blocks", m_active_blocks.m_abm_list.size());
+			g_profiler->avg("SEnv: active blocks cached", m_active_blocks.m_abm_list.size() - abm_not_cached);
```
Line is too long (> 80 characters)

lhofhansl replied:
> Why do you need the cache at all? A boolean value would do it too ...

Huh? That's the whole point of this :)
ABMs are per node-type, not per block. We also do not want to invalidate all blocks somehow when a new ABM is added. The way it is now the cache is only invalidated when the block is changed and on top the decision is block local. That makes this simple and safe.

SmallJoker replied:
Alright, I didn't even expect that the ABMs are processed from scratch each second - thought they were a static vector of ABM trigger information per content ID.

lhofhansl replied:
@Cazadorro we can argue about the naming of the variables. :)
But please have a look at the code where they are used.

lhofhansl replied:
@SmallJoker There're definitely more things to do. For example we could refresh the list of ABMs less frequently than we trigger them. In that case we could cache everything in a bit (as you suggest) until we refresh ABMs the next time.

Perhaps ABMs are not usually added often. So invalidating everything in that case might be OK. That would involve more complex logic with timestamps.

Cazadorro commented:
What is ABM? what does it stand for?  Also why is `do_not_cache_contents == false` not `cache_contents = true`?

Cazadorro commented:
@rubenwardy  In addition to the forums and direct links to dev wiki being blocked, [I didn't realize this was domain specific to minetest]( http://lmgtfy.com/?q=abm+c%2B%2B), my bad.

lhofhansl commented:
@Cazadorro No problem. Thanks for taking a look.

nerzhul commented on src/serverenvironment.cpp:
```diff
@@ -818,6 +838,15 @@ class ABMHandler
+			// cache content types as we go
+			if (!block->contents_cached && !block->do_not_cache_contents) {
+				block->contents.insert(c);
+				if (block->contents.size() > 64) {
```
a thread local g_setting can be nice to use here

paramat replied:
I don't see a need for a setting.

sfan5 replied:
Why not? If you have lots of RAM to spare, this would be one of the many small adjustments you could make to improve the server's responsiveness.

lhofhansl replied:
Nah. This is just a heuristic, as long as node types for "most" blocks can be cached this is good.
A block can have at most 16^3 = 4096 different node types. Caching at most 1/64th seems good.

1. If we cache too much there's additional memory *and* additional CPU to check the cache first
2. The more different node types we have the more likely it becomes that ABMs need to be run for this block. And this check is entirely in vain.

16, 32, or 64 are good. Let's not add another config option to MT, it has *way* too many already.

paramat commented on src/serverenvironment.cpp:
```diff
@@ -800,11 +800,31 @@ class ABMHandler
 		if(m_aabms.empty() || block->isDummy())
 			return;
 
+		// check the content type cache first
```
'Check'

paramat commented on src/serverenvironment.cpp:
```diff
@@ -800,11 +800,31 @@ class ABMHandler
 
+		// check the content type cache first
+		// to see whether there are any ABMs
+		// to be run at all for this block
```
Full stop at end of line.

paramat commented on src/mapblock.h:
```diff
@@ -529,6 +531,14 @@ class MapBlock
 	static const u32 nodecount = MAP_BLOCKSIZE * MAP_BLOCKSIZE * MAP_BLOCKSIZE;
 
+	//// ABM optimizations ////
+	// cache of content types
```
For all comments: Capital letter at start.

paramat commented on src/serverenvironment.cpp:
```diff
@@ -818,6 +838,15 @@ class ABMHandler
+			if (!block->contents_cached && !block->do_not_cache_contents) {
+				block->contents.insert(c);
+				if (block->contents.size() > 64) {
+					// too many different nodes... don't try to cache
```
'Too'

lhofhansl replied:
It says "too". You mean upper case T?
It's not a full sentence, though.

paramat replied:
Yes capital 'T' please.

paramat commented:
I studied the ABM code recently and this seems good, i'm close to approving.

lhofhansl commented:
Thanks!
25cc5d1a3263aaca46f77f64b6cc20b1d2c32cac
